File: src/main/java/me/mattstudios/mattcore/locale/Locale.java
package me.mattstudios.mattcore.locale;

import ch.jalu.configme.SettingsHolder;
import ch.jalu.configme.SettingsManager;
import ch.jalu.configme.properties.Property;
import org.bukkit.Bukkit;
import org.bukkit.command.CommandSender;
import org.bukkit.plugin.Plugin;

import java.io.File;

import static ch.jalu.configme.SettingsManagerBuilder.withYamlFile;
import static ch.jalu.configme.configurationdata.ConfigurationDataBuilder.createConfiguration;
import static me.mattstudios.mattcore.utils.MessageUtils.color;

@SuppressWarnings("unused")
public class Locale {

    // The plugin's instance
    private final Plugin plugin;
    // The settings holder of the languages
    private SettingsManager settingsManager;

    /**
     * Constructor of the locale
     *
     * @param plugin The plugin instance
     */
    public Locale(Plugin plugin) {
        this.plugin = plugin;
    }

    /**
     * Loads the messages file
     *
     * @param messageClass The message file class
     * @param fileName     The file name
     */
    public void load(final Class<? extends SettingsHolder> messageClass, final String fileName) {
        settingsManager = withYamlFile(
                new File(plugin.getDataFolder(), "lang/" + fileName + ".yml"))
                .useDefaultMigrationService()
                .configurationData(createConfiguration(messageClass))
                .create();
    }

    /**
     * Gets a message from the file colored
     *
     * @param property The property path for the message
     * @return The message already colored
     */
    public String getMessage(final Property<String> property) {
        return color(settingsManager.getProperty(property));
    }

    /**
     * Gets a message from the file raw
     *
     * @param property The property path for the message
     * @return The message already colored
     */
    public String getMessageRaw(final Property<String> property) {
        return settingsManager.getProperty(property);
    }

    /**
     * Sends a message to the sender from the lang file
     *
     * @param sender   The sender
     * @param property The property
     */
    public void sendMessage(final CommandSender sender, final Property<String> property) {
        sender.sendMessage(color(settingsManager.getProperty(property)));
    }

    /**
     * Sends a message to the sender from the lang file
     *
     * @param property The property
     */
    public void sendMessage(final Property<String> property) {
        Bukkit.getConsoleSender().sendMessage(color(settingsManager.getProperty(property)));
    }

    /**
     * Used to reload the lang file
     */
    public void reload() {
        settingsManager.reload();
    }

}

File: src/main/kotlin/me/mattstudios/core/locale/Locale.kt
package me.mattstudios.core.locale

import me.mattstudios.config.SettingsHolder
import me.mattstudios.config.SettingsManager
import me.mattstudios.config.properties.Property
import java.io.File
import java.nio.file.Paths

/**
 * **Locale** handler for all plugins
 */
@Suppress("MemberVisibilityCanBePrivate")
class Locale(private val folder: File) {

    // Current being used language
    var language = Language.ENGLISH
        private set

    // The settings holder of the languages
    private lateinit var settingsManager: SettingsManager

    // The message [SettingHolder]
    private lateinit var messageClass: Class<out SettingsHolder>

    /**
     * Sets the message [SettingsHolder] to be used for the [Locale]
     */
    fun setHolder(messageClass: Class<out SettingsHolder>): Locale {
        this.messageClass = messageClass
        return this
    }

    /**
     * Sets the current [Language]
     */
    fun setLocale(language: Language): Locale {
        this.language = language
        return this
    }

    /**
     * Creates the [SettingsManager]
     */
    fun create() {
        settingsManager = SettingsManager.from(Paths.get(folder.path, "lang/${language.file}.yml"))
            .configurationData(messageClass)
            .create()
    }

    /**
     * Gets a message from the locale file
     */
    operator fun <T> get(property: Property<T>) = settingsManager[property]

    /**
     * Used to reload the lang file
     */
    fun reload() = settingsManager.reload()

}